connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java

```diff
@@ -0,0 +1,525 @@
+        // and fail the task once unpaused
+        transactionManager.maybeCommitFinalTransaction();
+    }
+
+    private void maybeBeginTransaction() {
+        if (!transactionOpen) {
+            producer.beginTransaction();
+            transactionOpen = true;
+        }
+    }
+
+    private void commitTransaction() {
+        log.debug("{} Committing offsets", this);
+
+        long started = time.milliseconds();
+
+        // We might have just aborted a transaction, in which case we'll have to begin a new one
+        // in order to commit offsets
+        maybeBeginTransaction();
+
+        AtomicReference<Throwable> flushError = new AtomicReference<>();
+        Future<Void> offsetFlush = null;
+        if (offsetWriter.beginFlush()) {
+            // Now we can actually write the offsets to the internal topic.
+            offsetFlush = offsetWriter.doFlush((error, result) -> {
+                if (error != null) {
+                    log.error("{} Failed to flush offsets to storage: ", ExactlyOnceWorkerSourceTask.this, error);
+                    flushError.compareAndSet(null, error);
+                } else {
+                    log.trace("{} Finished flushing offsets to storage", ExactlyOnceWorkerSourceTask.this);
+                }
+            });
+        }
+
+        // Commit the transaction
+        // Blocks until all outstanding records have been sent and ack'd
+        try {
+            producer.commitTransaction();
+            if (offsetFlush != null) {
+                // Although it's guaranteed by the above call to Producer::commitTransaction that all outstanding
+                // records for the task's producer (including those sent to the offsets topic) have been delivered and
+                // ack'd, there is no guarantee that the producer callbacks for those records have been completed. So,
+                // we add this call to Future::get to ensure that these callbacks are invoked successfully before
+                // proceeding.
+                offsetFlush.get();
+            }
+        } catch (Throwable t) {
+            flushError.compareAndSet(null, t);
+        }
+
+        transactionOpen = false;
+
+        Throwable error = flushError.get();
+        if (error != null) {
+            recordCommitFailure(time.milliseconds() - started, null);
+            offsetWriter.cancelFlush();
+            throw maybeWrapProducerSendException(
+                    "Failed to flush offsets and/or records for task " + id,
+                    error
+            );
+        }
+
+        transactionMetrics.commitTransaction();
+
+        long durationMillis = time.milliseconds() - started;
+        recordCommitSuccess(durationMillis);
+        log.debug("{} Finished commitOffsets successfully in {} ms", this, durationMillis);
+
+        // No need for any synchronization here; all other accesses to this field take place in producer callbacks,
+        // which should all be completed by this point
```

Comment:
That's a fair point. Right now this is a result of the threading model of the producer, which blocks on the completion of all in-flight batches and their callbacks before sending a request to the broker to end a transaction. This is not technically guaranteed by the documented behavior for `Producer::commitTransaction`, though; the only guarantees we get from the [Javadocs](https://kafka.apache.org/32/javadoc/org/apache/kafka/clients/producer/KafkaProducer.html#commitTransaction()) are:
> This method will flush any unsent records before actually committing the transaction. Further, if any of the [send(ProducerRecord)](https://kafka.apache.org/32/javadoc/org/apache/kafka/clients/producer/KafkaProducer.html#send(org.apache.kafka.clients.producer.ProducerRecord)) calls which were part of the transaction hit irrecoverable errors, this method will throw the last received exception immediately and the transaction will not be committed.

If we don't want to rely on this technically-undocumented behavior, we can add some more bookkeeping to ensure this ourselves. Alternatively, we can rely on this behavior and I can remove the comment above about producer callback guarantees and the bookkeeping we already do on that front (which I've done for now).

@hachikuji @guozhangwang do you have thoughts about whether it should be part of the contract for `Producer::commitTransaction` that all user-supplied callbacks for records within the transaction are fired before the method returns? The use case is tracking record metadata that's only available after the record is ack'd and then reporting it to source tasks, before clearing a local cache of that metadata.

Reply:
@C0urante thanks for the explanation. I don't think a revised `Producer` contract would be sufficient though. It's not the _blocking_ per-se, it's the visibility of mutations to `commitableRecords` done on those threads being correctly synchronized with reads on this line. `Producer` can't help with that because it doesn't know about `commitableRecords`, and so the writes and the read could be re-ordered anyway.

Reply:
Ah, fair. I was taking a charitable interpretation of the [Javadocs](https://docs.oracle.com/javase/8/docs/api/java/util/LinkedHashMap.html), which state:

> If multiple threads access a linked hash map concurrently, and at least one of the threads modifies the map structurally, it must be synchronized externally.

In this case, by the time we get to this line, we don't technically have multiple threads accessing the map concurrently; it's guaranteed that this will be the only thread reading from or writing to the map. But it costs very little to add the synchronization barrier, so I've done that.

Also emailed Jason and Guozhang for their thoughts on callbacks with transactional producers. Hoping to hear back soon 🤞

Reply:
Yeah, I think it's a documentation gap. It is intended for callbacks to be invoked before `EndTxn` is sent. It would be a little strange otherwise. Worth noting that the callback is just for notification. Even if the callback raises an exception, we won't stop an active commit.

Reply:
Thanks Jason, filed https://issues.apache.org/jira/browse/KAFKA-13967 to track the producer docs improvement and will have a PR out for it shortly.

Reply:
Javadocs PR: https://github.com/apache/kafka/pull/12264